### PR TITLE
TT-5776 Racy filterData func

### PR DESCRIPTION
**Pull request by tbuchaillot**
## Description

This PR changes the way we duplicate the slice while filtering. The previous way was more mem efficient but it was racy when using multiple pumps.

This PR also adds more tests scenarios for filtering data.

## Related Issue

https://tyktech.atlassian.net/browse/TT-5776
## Motivation and Context

https://tyktech.atlassian.net/browse/TT-5776
## How This Has Been Tested

Added unit test. 
Ran test with -race

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -299,10 +299,13 @@ func filterData(pump pumps.Pump, keys []interface{}) []interface{} {
 	if !filters.HasFilter() && !pump.GetOmitDetailedRecording() && !shouldTrim {
 		return keys
 	}
-	filteredKeys := keys[:]
+
+	filteredKeys := make([]interface{}, len(keys))
```
Comment (tbuchaillot) [inline main.go:303]: this is the brain of the fix. In concurrent scenarios - and with filters, when modifying `filteredKeys` we are modifying the `keys` slice.

With this solution, we are copying all the keys in this new slice (only if it has filters or omit_detailed_recording) in this way we don't override the original slice.

Comment (buger) [reply main.go:303]: I wonder if for performance reasons we should do this copy ONLY if we have at least one filter?

Comment (tbuchaillot) [reply main.go:303]: @buger that's already happening.

we return the same slice without copying it if it doesn't have filters / or need some general sanitization https://github.com/TykTechnologies/tyk-pump/blob/027a664557c7f095fef3383614bc8b60619fa520/main.go#L299

**Review by buger — APPROVED** [review]
Makes sense!

Love the tests!

**Pull request merged**